### PR TITLE
Replace bcrypto with standalone libraries

### PR DESCRIPTION
**Pull request by mpetrun5**
Removes bcrypto from dependencies.

Replaced with: 
 - [futoin-hdkf](https://www.npmjs.com/package/futoin-hkdf)
 - [aead-js](https://www.npmjs.com/package/aead-js)
 - [tweetnacl](https://www.npmjs.com/package/tweetnacl)

Related to issue: #27 

cc: @mpetrunic

### CODE REVIEW

File: src/handshakes/abstract-handshake.ts [hunk 1]
```diff
@@ -104,10 +106,7 @@ export abstract class AbstractHandshake {
 
   protected dh(privateKey: bytes32, publicKey: bytes32): bytes32 {
     try {
-      const derived = x25519.derive(publicKey, privateKey);
-      const result = Buffer.alloc(32);
-      derived.copy(result);
-      return result;
+      return Buffer.from(box.before(publicKey, privateKey), 0, 32);
```
Comment (mpetrunic) [inline src/handshakes/abstract-handshake.ts:109]: What are you trying to achieve here?
Buffer view over 0...32 byte of output?

File: src/handshakes/abstract-handshake.ts [hunk 2]
```diff
@@ -119,7 +118,9 @@ export abstract class AbstractHandshake {
   }
 
   protected getHash(a: bytes, b: bytes): bytes32 {
-    return SHA256.digest(Buffer.from([...a, ...b]));
+    return Buffer.from(
+      hash.sha256().update(Buffer.from([...a, ...b])).digest('hex')
```
Comment (mpetrunic) [inline src/handshakes/abstract-handshake.ts:122]: This hash is invalid. Buffer.from uses utf-8 encoding by default.

To reduce memory allocation:
```javascript
const hash = hash.sha256().update(Buffer.from([...a, ...b])).digest();
Buffer.from(hash.buffer, hash.offset, hash.length);
```
https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_arraybuffer_byteoffset_length

File: src/utils.ts [hunk 3]
```diff
@@ -111,5 +111,9 @@ export function getHkdf(ck: bytes32, ikm: bytes): Hkdf {
 }
 
 export function isValidPublicKey(pk: bytes): boolean {
-  return x25519.publicKeyVerify(pk.slice(0, 32));
+  if(pk.length !== 32 || pk.equals(Buffer.alloc(32))){
+    return false;
+  }
+
+  return true;
```
Comment (mpetrunic) [inline src/utils.ts:118]: @morrigan Any idea how to implement proper public key verification?

File: test/noise.test.ts [hunk 4]
```diff
@@ -3,8 +3,8 @@ import DuplexPair from 'it-pair/duplex';
 import {Noise} from "../src";
 import {createPeerIdsFromFixtures} from "./fixtures/peer";
 import Wrap from "it-pb-rpc";
-import {random} from "bcrypto";
 import sinon from "sinon";
+import crypto from 'crypto';
```
Comment (mpetrunic) [inline test/noise.test.ts:7]: You didn't list this dependency.
Also you have this method in libp2p-crypto
https://github.com/libp2p/js-libp2p-crypto#cryptorandomBytesnumber

File: src/utils.ts [hunk 5]
```diff
@@ -10,8 +11,9 @@ import {pb} from "./proto/payload";
 const NoiseHandshakePayloadProto = pb.NoiseHandshakePayload;
 
 export function generateKeypair(): KeyPair {
-  const privateKey = x25519.privateKeyGenerate();
-  const publicKey = x25519.publicKeyCreate(privateKey);
+  const keyPair = box.keyPair();
+  const publicKey = Buffer.from(keyPair['publicKey']);
+  const privateKey = Buffer.from(keyPair['secretKey']);
```
Comment (mpetrunic) [inline src/utils.ts:16]: any reason why you didn't do `keyPair.publicKey`?

File: src/handshakes/abstract-handshake.ts (hunk 1 above)
Comment (mpetrun5) [reply src/handshakes/abstract-handshake.ts]: I unnecessarily tried to replicate the old logic.

File: src/handshakes/abstract-handshake.ts [hunk 6]
```diff
@@ -106,7 +106,8 @@ export abstract class AbstractHandshake {
 
   protected dh(privateKey: bytes32, publicKey: bytes32): bytes32 {
     try {
-      return Buffer.from(box.before(publicKey, privateKey), 0, 32);
+      const sharedKey = box.before(publicKey, privateKey)
+      return Buffer.from(sharedKey);
```
Comment (mpetrunic) [inline src/handshakes/abstract-handshake.ts:110]:
```suggestion
      return Buffer.from(sharedKey.buffer, sharedKey.offset, sharedKey.length);
```
It won't allocate new array

File: src/handshakes/abstract-handshake.ts [hunk 7]
```diff
@@ -118,9 +119,8 @@ export abstract class AbstractHandshake {
   }
 
   protected getHash(a: bytes, b: bytes): bytes32 {
-    return Buffer.from(
-      hash.sha256().update(Buffer.from([...a, ...b])).digest('hex')
-    );
+    const hashValue = hash.sha256().update(Buffer.from([...a, ...b])).digest();
+    return Buffer.from(hashValue);
```
Comment (mpetrunic) [inline src/handshakes/abstract-handshake.ts:123]: same as above

Comment (mpetrun5) [reply src/handshakes/abstract-handshake.ts:123]: hashValue is number[]. it doesn't have buffer or offset.

Comment (mpetrunic) [reply src/handshakes/abstract-handshake.ts:123]: Could you try it. It's probably Uint8Array but it's displayed as number[] in logs
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array

Comment (mpetrunic) [reply src/handshakes/abstract-handshake.ts:123]: nvm, it's regular Array, wierd. I guess it's fine like this

File: src/utils.ts (hunk 3 above)
Comment (morrigan) [reply src/utils.ts:118]: Can't find anything. I guess no big deal as it won't pass DH otherwise

**Comment by morrigan** [comment]
What's the size now?

**Comment by mpetrun5** [comment]
image
Webpack analyzer is kinda broken in this image, but It says 1020 kib.
https://bundlephobia.com/scan-results?packages=@types/bn.js@4.11.6,aead-js@0.1.0,bn.js@5.0.0,buffer@5.4.3,debug@4.1.1,futoin-hkdf@1.3.2,hash.js@1.1.7,it-buffer@0.1.1,it-length-prefixed@3.0.0,it-pair@1.0.0,it-pb-rpc@0.1.8,it-pipe@1.1.0,libp2p-crypto@0.17.1,peer-id@0.13.5,protobufjs@6.8.8,tweetnacl@1.0.3

It is not much better.

**Comment by mpetrunic** [comment]
I wonder if we can reduce amount of stuff imported from libp2p-crypto since that is pulling node-forge. I guess that something guys from libp2p will do

**Comment by morrigan** [comment]
Surprisingly, it's not even better than #47 :/

File: test/noise.test.ts [hunk 8]
```diff
@@ -3,8 +3,8 @@ import DuplexPair from 'it-pair/duplex';
 import {Noise} from "../src";
 import {createPeerIdsFromFixtures} from "./fixtures/peer";
 import Wrap from "it-pb-rpc";
-import {random} from "bcrypto";
 import sinon from "sinon";
+import crypto from 'libp2p-crypto';
```
Comment (mpetrunic) [inline test/noise.test.ts:7]: try to
`import {randomBytes} from "libp2p-crypto"`

**Comment by mpetrun5** [comment]
It shows no difference in webpack-bundle-analyzer.

**Comment by mpetrunic** [comment]
@morrigan Can you re-review it, it somewhat smaller now, but I don't think it can get better without fixing js-libp2p-crypto. That said not sure if we should just apply last changes in #47 or should we merge this?

**Comment by mpetrunic** [comment]
I will make a release out of it and try it in interop repo with tests against go version

**Pull request merged**